cct/module.py

```diff
@@ -97,21 +104,53 @@ def check_module_py(self, candidate):
             # Check that class is from our namespace
             if module_name == clazz.__module__:
                 # Instantiate class
+                self.check_module_version(name)
                 cls = getattr(module, name)
                 if issubclass(cls, Module):
-                    self.modules[module_name.split('.')[-1] + "." + cls.__name__] = cls(module_name.split('.')[-1] + "." + cls.__name__, os.path.dirname(candidate), os.path.join(self.artifacts_dir, name.split('.')[0]))
+                    name = module_name.split('.')[-1] + "." + cls.__name__
+                    self.modules[name] = cls(name, os.path.dirname(candidate), os.path.join(self.artifacts_dir, name.split('.')[0]))
+                    self.modules[name].version = self.version
+                    self.modules[name].override = self.override
 
     def check_module_sh(self, candidate):
         module_name = "cct.module." + os.path.dirname(candidate).split('/')[-1]
-        logger.debug("importing module %s to %s" % (os.path.abspath(candidate), module_name))
+        logger.debug("Importing module %s to %s" % (os.path.abspath(candidate), module_name))
         name = module_name.split('.')[-1] + "." + os.path.basename(candidate)[:-3]
+        if self.check_module_version(name):
+            return
         self.modules[name] = ShellModule(name, os.path.dirname(candidate),
                                          os.path.join(self.artifacts_dir, name.split('.')[0]), candidate)
+        self.modules[name].version = self.version
+        self.modules[name].override = self.override
+
+    def check_module_version(self, name):
+        # if we override version we dont care about overwriting modules
+        if self.override:
+            return False
+        # if the module doesnt exists just create it
+        if name not in self.modules:
+            return False
+        # if module was froced we will not override it
+        if self.modules[name].override:
+            return True
+        # check if the version are not conflicting
+        # FIXME - tag in branch vs branch is considered incompatible
+        if self.modules[name].version != self.version:
+            logger.error("Conflicting module '%s' version found, installed version '%s', required version '%s'" % (name, self.modules[name].version, self.version))
+            raise Exception('Conflicting modules')
+        return True
 
     def list(self):
         print("available cct modules:")
-        for module, _ in self.modules.iteritems():
-            print("  %s" % module)
+        for name, module in self.modules.iteritems():
+            if name == 'base.Dummy':
```

Comment:
dummy is special module used a lot in test - It needs to be moved to better place in future